### PR TITLE
Added new extraction syntax.

### PR DESCRIPTION
**Pull request by ruschein**

### CODE REVIEW

File: cpp/convert_json_to_marc.cc [hunk 1]
```diff
@@ -101,17 +104,31 @@ void LoadISSNsToJournalTitlesAndPPNsMap(std::unordered_map<std::string, JournalT
             LOG_ERROR("malformed line #" + std::to_string(line_no) + " in \"" + MAP_FILE_PATH + "\"!");
         issns_to_journal_titles_and_ppns_map->emplace(parts[0], JournalTitleAndPPN(parts[1], parts[2]));
     }
+
+    LOG_INFO("Loaded " + std::to_string(issns_to_journal_titles_and_ppns_map->size())
+             + " mappings from print ISSN's to online ISSN's, PPPN's and journal titles.");
```
Comment (mtrojan-ub) [inline cpp/convert_json_to_marc.cc:109]: Why "ppn" in variable name but "pppn" in log string?

File: cpp/convert_json_to_marc.cc [hunk 2]
```diff
@@ -323,9 +343,11 @@ std::vector<FieldDescriptor> LoadFieldDescriptors(const std::string &inifile_pat
             if (tag.length() != MARC::Record::TAG_LENGTH)
                 LOG_ERROR("invalid tag \"" + tag + "\" in section \"" + section_name + "\" in \"" + ini_file.getFilename() + "\"!");
 
-            std::vector<std::pair<char, std::string>> subfield_codes_to_json_tags, subfield_codes_to_prefixes, subfield_codes_to_fixed_subfields;
+            std::vector<std::pair<char, std::string>> subfield_codes_to_json_paths, subfield_codes_to_prefixes, subfield_codes_to_fixed_subfields;
             std::map<char, std::shared_ptr<RegexMatcher>> subfield_codes_to_extraction_regexes_map;
+            std::vector<DependentSubfield> dependent_subfields;
             for (const auto section_entry : section) {
+if (section_entry.name_ == "Host Item Entry")LOG_ERROR("Found Host Item Entry");
```
Comment (mtrojan-ub) [inline cpp/convert_json_to_marc.cc:350]: This seems to be for debugging and should be removed

**Comment by ruschein** [comment]
removed


--
Dr. Johannes Ruscheinski
Universitätsbibliothek Tübingen - IT-Abteilung -
Wilhelmstr. 32, 72074 Tübingen

Tel: +49 7071 29-72820
FAX: +49 7071 29-5069
Email: johannes.ruscheinski@uni-tuebingen.de

 The Sophisticate:  "The world isn't black and white.  No one does pure good or pure bad. It's all gray.  Therefore, no one is better than anyone else."
    The Zetet:  "Knowing only gray, you conclude that all grays are the same shade.  You mock the simplicity of the two-color view, yet you replace it with a one-color view..."
      —Marc Stiegler, David's Sling


________________________________

File: cpp/data/core.conf [hunk 3]
```diff
@@ -72,3 +72,12 @@ map_to_marc_language_code   = true
 [Publisher]
 tag                         = 264
 subfield_b                  = publisher
+
+[Host Item Entry]
+tag                         = 773
+subfield_a                  = map:journal_tites_and_ppns_map{journals/0/identifiers/0}:title
```
Comment (mtrojan-ub) [inline cpp/data/core.conf:78]: tites => titles?

File: cpp/data/core.conf [hunk 4]
```diff
@@ -72,3 +72,12 @@ map_to_marc_language_code   = true
...
+subfield_a_extraction_regex = "\\d{4}-?\\d{3}[0-9xX]"
+subfield_w                  = map:journal_tites_and_ppns_map{journals/0/identifiers/0}:ppn
```
Comment (mtrojan-ub) [inline cpp/data/core.conf:80]: again: titles => titles?

File: cpp/convert_json_to_marc.cc [hunk 5]
```diff
@@ -414,20 +443,77 @@ std::string ReferencedJSONDataStateToString(const ReferencedJSONDataState refere
...
-    for (const auto &subfield_code_and_json_tag : subfield_codes_to_json_tags) {
-        const auto node(object->deepResolveNode(subfield_code_and_json_tag.second));
+    for (const auto &subfield_code_and_json_path : subfield_codes_to_json_paths) {
+        std::string relative_json_path(subfield_code_and_json_path.second);
+
+        // Two types of relative JSON path references exist: 1) the direct one and
+        // 2) one nested within a "map:" expression.
+        std::string map_name, json_path, map_key;
+        bool use_first_array_element_only;
+        if (SplitMapReference(subfield_code_and_json_path.second, &map_name, &json_path, &map_key,
```
Comment (mtrojan-ub) [inline cpp/convert_json_to_marc.cc:500]: Shouldn't map_name and map_key be used somehow?

Comment (ruschein) [reply cpp/convert_json_to_marc.cc:500]: Not here.  This is just for a preliminary check of the type of JSON references.  At other call sites of SplitMapReference, map_name and map_key are indeed used.

**Pull request merged**